crates/ironrdp-rdpdr-native/src/nix/simple.rs

```diff
@@ -0,0 +1,790 @@
+                            },
+                        ))])
+                    }
+                }
+                Err(error) => {
+                    warn!(%error, "Get metadata error");
+                    Ok(vec![SvcMessage::from(RdpdrPdu::ClientDriveQueryDirectoryResponse(
+                        ClientDriveQueryDirectoryResponse {
+                            device_io_reply: DeviceIoResponse::new(device_io_request, not_found_status),
+                            buffer: None,
+                        },
+                    ))])
+                }
+            }
+        }
+    }
+}
+pub(crate) fn query_directory(
+    backend: &mut SimpleRdpdrBackend,
+    req_inner: ServerDriveQueryDirectoryRequest,
+) -> PduResult<Vec<SvcMessage>> {
+    match backend.file_path_map.get(&req_inner.device_io_request.file_id) {
+        Some(parent_pos_for_next) => {
+            let mut find_file_name = None;
+            if req_inner.initial_query > 0 {
+                if req_inner.path.ends_with('*') {
+                    let mut parent = backend.file_base.clone();
+                    let query_path = req_inner.path.replace('\\', "/");
+                    let len = query_path.len();
+                    // path ends with *, so its len > 0
+                    #[allow(clippy::arithmetic_side_effects)]
+                    parent.push_str(&query_path[0..len - 1]);
+                    let dirp = if let Ok(real_parent_c) = CString::new(parent.clone()) {
+                        // SAFETY: it is safe to call this on *nix platform
+                        unsafe { opendir(real_parent_c.as_ptr()) }
+                    } else {
+                        std::ptr::null::<DIR>().cast_mut()
+                    };
+                    if !dirp.is_null() {
+                        loop {
+                            // SAFETY: it is safe to call readdir on *nix platform
+                            let first = unsafe { readdir(dirp) };
+                            if first.is_null() {
+                                break;
+                            }
+                            // SAFETY: it is safe to call this for rdp
+                            let dir_name = unsafe { &(*first).d_name };
+                            // SAFETY: it is safe to call this for rdp
+                            let dir_name_in_rs =
+                                unsafe { CStr::from_ptr(dir_name as *const i8).to_string_lossy().into_owned() };
+                            if dir_name_in_rs == "." || dir_name_in_rs == ".." {
+                                continue;
+                            }
+                            parent.push_str(&dir_name_in_rs);
+                            find_file_name = Some(parent);
+                            break;
+                        }
+                        backend
+                            .file_dir_map
+                            .insert(req_inner.device_io_request.file_id, DirBox { dir: dirp });
```

Comment:
question: What happens when we keep around a DirBox with a null pointer? Is it intended? Shouldn’t we use an `Option<DirBox>` instead? Do we need to keep track of this at all when it failed?